app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts

```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
   container.appendChild(div)
 }
 
+/**
+ * when using Record notation, the following error is thrown:
+ * 'TS2456: Type alias 'BodyValue' circularly references itself.'
+ */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+type BodyValue = boolean | number | string | { [key: string]: BodyValue }
+
+/**
+ * Transforms an object into form data representation, also URL-encoding the values,
+ * Example:
+ * {
+ *   a_string: 'hello',
+ *   an_array: [
+ *     { first: 1 },
+ *     { second: 1, extra_param: 'with whitespace'}
+ *   ]
+ * }
+ * =>
+ * {
+ *   a_string: 'hello',
+ *   'an_array[0][first]': '1',
+ *   'an_array[1][second]': '1',
+ *   'an_array[1][extra_param]': 'with%20whitespace'
+ * }
+ * @param object
+ */
+export const objectToFormData = (object: BodyValue): Record<string, boolean | number | string> => {
+  const buildFormData = (formData: Record<string, boolean | number | string>, data: BodyValue, parentKey?: string) => {
+    if (data && typeof data === 'object') {
+      Object.keys(data).forEach((key: string) => {
+        buildFormData(formData, data[key], parentKey ? `${parentKey}[${key}]` : key)
+      })
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
```

Comment:
Several issues with this method:
I always find it very confusing two mix up declaration order, object comes as an argument and it's should be the focus of this function but instead is used at the very end for the first time. `buildFormData` is complex enough to be declared elsewhere, in fact it can be extracted without refactoring since it does not depend on `objectToFormData`.

The fact that an empty object is being populated from scratch suggests to me a `reduce` could be used here, ideally:
```js
formData = body.reduce(buildFormData, {})
```

Recurrence is nice provided the method signature doesn't change but here both `data` and `parentKey` generate control couples which IMHO means it should be split (at least) in two. On top of that `formData` is empty during the first iteration colliding with the actual method signature:
```ts
Record<string, never>
// vs
Record<string, boolean | number | string>
```

There is not one test for this method. I've spent a fair amount of time looking at it and still cannot get the grasp of it so it will be hard to maintain. Considering its purpose is pretty straightforward I strongly suggest it be refactored. I'm working on an alternative right now, I really rather sacrifice verbosity for readability.

Reply:
I think `buildFormData` is OK to be inside `objectToFormData`. Otherwise the only thing `objectToFormData` would do is to call `buildFormData`, no need to create a new function for that. In any case, what I would do is remove `objectToFormData` and call `buildFormData({}, body)` directly from `transformReportRequestBody`.

Reply:
> The fact that an empty object is being populated from scratch suggests to me a reduce could be used here

That's what I thought initially, but the problem is that we do need to traverse the whole object tree, so I am not sure that `reduce` would be that helpful (or easy to implement).

I'll address the rest of the comments later.

Reply:
> I always find it very confusing two mix up declaration order, object comes as an argument and it's should be the focus of this function but instead is used at the very end for the first time.

Well, the only reason why it is used in the end is because `buildFormData` is declared inside. I can extract it from there, no problem. It's just that I think thought of `objectToFormData` as a function that does everything, I think it's pretty useless if we just keep the last 3 lines (basically, I agree with what @jlledom says, except I don't think we need to call `buildFormData({}, body)` directly from `transformReportRequestBody`.
Anyway, I will see if I can make it more readable.

> `buildFormData` is complex enough to be declared elsewhere, in fact it can be extracted without refactoring since it does not depend on objectToFormData.

But do we need to have an exported function somewhere else (where?) if we know it is not going to be used by anything rather than this function? To me it makes more sense to keep the things together in this case, this will avoid the need to jump around multiple files/lines in the same file to understand what the function does.

> On top of that `formData` is empty during the first iteration colliding with the actual method signature: [...]

Doesn't `Record<string, boolean | number | string>` accept `{}` ?

> There is not one test for this method.

Fair enough, although the test would be for `objectToFormData` as it's the one that is exported.

> I've spent a fair amount of time looking at it and still cannot get the grasp of it so it will be hard to maintain.

What's confusing about it? What would you suggest?

Reply:
I tried to give it a spin, but I did not manage to get rid of recursion in favor of `reduce` and friends.

I don't think it makes sense to waste more time on this. It's a very specific case, and it currently works for the issue it is meant to solve.

If you have an alternative implementation - please share :wink:

Reply:
I won't insist, I lost this battle. One last war cry in favor of SRP though: a function that _does everything_ is *wrong*. A function must do one and only one thing. And it's OK to have a 1 or 2 line function as long as it complies with SRP. It makes it more readable and more maintainable.

Reply:
LOL, "function that does everything"  was probably an overstatement :sweat_smile: IMO `objectToFormData` actually does 1 thing (transform one object format to another), but it's self-contained, so you don't need to jump around the code to understand what it does.

But well, I understand your point also. Does this look better to you? https://github.com/3scale/porta/pull/3294/commits/a7a2e450b11231ffed306df167dda82459d9b1d0